text/0617-cloudfront-oac-l2.md

```diff
@@ -0,0 +1,648 @@
+    class Import extends OriginAccessControlBase {
+      public readonly originAccessControlId = originAccessControlId;
+      constructor(s: Construct, i: string) {
+        super(s, i);
+
+        this.originAccessControlId = originAccessControlId;
+      }
+    }
+    return new Import(scope, id);
+  }
+
+  public readonly originAccessControlId: string;
+  constructor(scope: Construct, id: string, props: OriginAccessControlProps = {}) {
+    super(scope, id);
+
+    const resource = new CfnOriginAccessControl(this, 'Resource', {
+      originAccessControlConfig: {
+        description: props.description,
+        name: props.originAccessControlName ?? this.generateName(),
+        signingBehavior: props.signingBehavior ?? SigningBehavior.ALWAYS,
+        signingProtocol: props.signingProtocol ?? SigningProtocol.SIGV4,
+        originAccessControlOriginType: props.originAccessControlOriginType ?? OriginAccessControlOriginType.S3,
+      },
+    });
+
+    this.originAccessControlId = resource.attrId;
+  }
+
+  private generateName(): string {
+    const name = Stack.of(this).region + Names.uniqueId(this);
+    if (name.length > 64) {
+      return name.substring(0, 32) + name.substring(name.length - 32);
+    }
+    return name;
+  }
+
+}
+```
+
+#### New `S3BucketOacOrigin` class
+
+A new class `S3BucketOacOrigin` will implement the `bind()` method to setup the OAC and update the bucket policy.
+
+In the case where an imported bucket is being used for the S3 origin, calling `bucket.addToResourcePolicy()` will fail to add the policy statement. Existing
+[workarounds](https://github.com/aws/aws-cdk/issues/6548#issuecomment-869091553) require the user to create a new `BucketPolicy` for the bucket and
+add the policy statements using `bucketPolicy.document.addStatements()`.
+However, this overwrites the whole bucket policy instead of appending statements to the
+existing policy which is a subpar user experience. The proposed solution to this issue is
+to use a custom resource to retrieve the existing bucket policy and append the
+OAC policy statement via the `GetBucketPolicy()` and `PutBucketPolicy()` API calls
+after the CloudFront distribution has been created.
+
+In the case where the S3 bucket uses SSE-KMS encryption (customer-managed key),
+a circular dependency error occurs when trying to deploy the template. When granting
+the CloudFront distribution access to use the KMS Key, there is a circular dependency:
+
+- CloudFront distribution references the S3 bucket
+- S3 bucket references the KMS key
+- KMS Key references the CloudFront distribution
+
+The proposed solution to this issue is to use a custom resource
+to retrieve and update the KMS key policy after the CloudFront
+distribution has been created via the `GetKeyPolicy()` and `PutKeyPolicy()` API calls.
+
+```ts
+class S3BucketOacOrigin extends cloudfront.OriginBase {
```

Comment:
I think at this point we shouldn't add a new origin class. Instead, what if we use `S3BucketOrigin` as an abstract class that defines a construct for an origin identity class (for existing behavior) and a origin access class using your implementation. This could look like this:

```ts
abstract class S3BucketOrigin extends cloudfront.OriginBase {
  public static withAccessIdentity(bucket: s3.IBucket, props: S3OriginProps = {}): S3BucketOrigin {
    return new (class OriginAccessIdentity extends S3BucketOrigin {
      private originAccessIdentity?: cloudfront.IOriginAccessIdentity;

      public constructor() {
        super(bucket, props);
      }

      public bind(scope: Construct, options: cloudfront.OriginBindOptions): cloudfront.OriginBindConfig {
        if (!this.originAccessIdentity) {
          // Using a bucket from another stack creates a cyclic reference with
          // the bucket taking a dependency on the generated S3CanonicalUserId for the grant principal,
          // and the distribution having a dependency on the bucket's domain name.
          // Fix this by parenting the OAI in the bucket's stack when cross-stack usage is detected.
          const bucketStack = cdk.Stack.of(this.bucket);
          const bucketInDifferentStack = bucketStack !== cdk.Stack.of(scope);
          const oaiScope = bucketInDifferentStack ? bucketStack : scope;
          const oaiId = bucketInDifferentStack ? `${cdk.Names.uniqueId(scope)}S3Origin` : 'S3Origin';

          this.originAccessIdentity = new cloudfront.OriginAccessIdentity(oaiScope, oaiId, {
            comment: `Identity for ${options.originId}`,
          });
        }
        // Used rather than `grantRead` because `grantRead` will grant overly-permissive policies.
        // Only GetObject is needed to retrieve objects for the distribution.
        // This also excludes KMS permissions; currently, OAI only supports SSE-S3 for buckets.
        // Source: https://aws.amazon.com/blogs/networking-and-content-delivery/serving-sse-kms-encrypted-content-from-s3-using-cloudfront/
        this.bucket.addToResourcePolicy(new iam.PolicyStatement({
          resources: [this.bucket.arnForObjects('*')],
          actions: ['s3:GetObject'],
          principals: [this.originAccessIdentity.grantPrincipal],
        }));
        return this._bind(scope, options);
      }

      protected renderS3OriginConfig(): cloudfront.CfnDistribution.S3OriginConfigProperty | undefined {
        if (!this.originAccessIdentity) {
          throw new Error('Origin access identity cannot be undefined');
        }
        return { originAccessIdentity: `origin-access-identity/cloudfront/${this.originAccessIdentity.originAccessIdentityId}` };
      }
    })();
  }

  public static withAccessControl(bucket: s3.IBucket, props: S3OriginProps = {}) {
    return new (class OriginAccessControl extends S3BucketOrigin {
      public constructor() {
        super(bucket, props);
      }

      // Add your origin access control implementation here ...
    })();
  }

  protected constructor(protected readonly bucket: s3.IBucket, props: S3OriginProps = {}) {
    super(bucket.bucketRegionalDomainName, props);
  }

  public abstract bind(scope: Construct, options: cloudfront.OriginBindOptions): cloudfront.OriginBindConfig;

  protected abstract renderS3OriginConfig(): cloudfront.CfnDistribution.S3OriginConfigProperty | undefined;

  protected _bind(scope: Construct, options: cloudfront.OriginBindOptions): cloudfront.OriginBindConfig {
    return super.bind(scope, options);
  }
}
```